src/Microsoft.AspNetCore.Authentication.OpenIdConnect/OpenIdConnectHandler.cs

```diff
@@ -684,20 +679,49 @@ private void PopulateSessionProperties(OpenIdConnectMessage message, Authenticat
         }
 
         /// <summary>
-        /// Redeems the authorization code for tokens at the token endpoint
+        /// Redeems the authorization code for tokens at the token endpoint.
+        /// 
+        /// The authorization response may contain error while its status code doesn't indicate failure. In this case
+        /// exception won't be thrown. The OpenIdConnectMessage will carry the error code.
         /// </summary>
         /// <param name="tokenEndpointRequest">The request that will be sent to the token endpoint and is available for customization.</param>
         /// <returns>OpenIdConnect message that has tokens inside it.</returns>
         protected virtual async Task<OpenIdConnectMessage> RedeemAuthorizationCodeAsync(OpenIdConnectMessage tokenEndpointRequest)
         {
             Logger.RedeemingCodeForTokens();
+
             var requestMessage = new HttpRequestMessage(HttpMethod.Post, _configuration.TokenEndpoint);
             requestMessage.Content = new FormUrlEncodedContent(tokenEndpointRequest.Parameters);
+
             var responseMessage = await Backchannel.SendAsync(requestMessage);
-            responseMessage.EnsureSuccessStatusCode();
-            var tokenResonse = await responseMessage.Content.ReadAsStringAsync();
-            var jsonTokenResponse = JObject.Parse(tokenResonse);
-            return new OpenIdConnectMessage(jsonTokenResponse);
+
+            var contentMediaType = responseMessage.Content.Headers.ContentType?.MediaType;
+            if (!string.IsNullOrEmpty(contentMediaType) && !string.Equals(contentMediaType, "application/json", StringComparison.OrdinalIgnoreCase))
+            {
+                Logger.LogWarning($"The token error response's content type is not application/json but {contentMediaType}.");
+            }
+
+            // Error handling:
+            // 1. If the response body can't be parsed as json, throws.
+            // 2. If the response's status code is not in 2XX range, throw OpenIdConnectProtocolException. If the body is correct parsed, 
+            //    pass the error information from body to the exception.
+            OpenIdConnectMessage message;
+            try
+            {
+                var responseContent = await responseMessage.Content.ReadAsStringAsync();
+                message = new OpenIdConnectMessage(responseContent);
+            }
+            catch (Exception ex)
```

Comment:
I'd rather not in case `OpenIdConnectMessage` change the internal json parsing logic.